api/api/src/main/scala/uk/ac/wellcome/platform/api/models/Aggregation.scala

```diff
@@ -69,87 +59,112 @@ object Aggregations extends Logging {
-        .recoverWith {
-          case err =>
-            warn("Failed to parse aggregation from ES", err)
-            Failure(err)
-        }
-        .toOption
-        .map(
-          agg =>
-            Aggregation(
-              agg.buckets
-                .map(getFilteredBucket)
-                .filter(_.count != 0)))
+  implicit val decodeGenreFromLabel: Decoder[Genre[Minted]] =
+    Decoder.decodeString.map { str =>
+      Genre(label = str)
+    }
+
+  implicit val decodeSubjectFromLabel: Decoder[Subject[Minted]] =
+    Decoder.decodeString.map { str =>
+      Subject(label = str, concepts = Nil)
     }
 
-    /** If the `filtered` sub-aggregation is present (see [[uk.ac.wellcome.platform.api.services.FiltersAndAggregationsBuilder]])
-      * then use the count from it */
-    private def getFilteredBucket[T](
-      esAggBucket: EsAggregationBucket[T]): AggregationBucket[T] =
-      AggregationBucket(
-        esAggBucket.key,
-        esAggBucket.filtered
-          .map(_.doc_count)
-          .getOrElse(esAggBucket.doc_count))
-
-    implicit def fromJsonAggReader[T: Decoder]: AggReader[T] =
-      (json: String) => fromJson[T](json)
+  implicit class EnhancedEsAggregations(
+    aggregations: com.sksamuel.elastic4s.requests.searches.Aggregations) {
+
+    def decodeAgg[T: Decoder](
+      name: String,
+      documentPath: Option[String] = None): Option[Aggregation[T]] = {
+      aggregations
+        .getAgg(name)
+        .flatMap(
+          _.safeTo[Aggregation[T]]((json: String) =>
+            AggregationMapping
+              .aggregationParser[T](json, documentPath)).recoverWith {
+            case err =>
+              warn("Failed to parse aggregation from ES", err)
+              Failure(err)
+          }.toOption
+        )
+    }
+  }
+}
+
+// This object contains the tools required to map an aggregation result from ES
+// into our Aggregation data type.
+//
+// There 2 independent variables in each bucket of these aggregations:
+//
+// ## Root count vs filtered count
+// If aggregation buckets have a subaggregation named `filtered` from [[uk.ac.wellcome.platform.api.services.FiltersAndAggregationsBuilder]]
+// then we use the count from there, otherwise we use the count from the root of the bucket.
+//
+// ## Key-only vs sample document
+// For some aggregation types `T` (in `Aggregation[T]`) we can decode `T` directly from the bucket key -
+// for example with WorkType, we can use `fromCode` to get the label of the WorkType given its id.
+// However, sometimes we need more information from the index, which is achieved using a top hits
+// aggregation of size 1 - see [[uk.ac.wellcome.platform.api.services.ElasticsearchQueryBuilder]].
+//
+// In this case there is a `sample_doc` subaggregation for which we can provide a
+// `path` to the relevant data inside it. For example, the Language type uses this to obtain both the
+// label and id at the path `data.language` within `sample_doc`.
```

Comment:
(BIKESHED WARNING!) I tend to use multiline comments for documentation purposes, and inline comments for describing implementation details. Also free text is IMO hard to read when over 80-100 chars, so might be good to limit the line length here.